File: Mud-pro-Backend/src/controllers/ugInventory/ugInventoryProductController.js
import UgInventorySnapshot from "../../modules/ugInventory/ugInventoryProductModel.js";

// ─────────────────────────────────────────────────────────────────
// GET  /api/ug-inventory/:wellId
// Returns the latest snapshot for a well (or empty defaults)
// ─────────────────────────────────────────────────────────────────
export const getUgInventory = async (req, res) => {
  try {
    const { wellId } = req.params;

    const snapshot = await UgInventorySnapshot.findOne({ wellId }).sort({
      updatedAt: -1,
    });

    if (!snapshot) {
      // Return empty structure so Flutter can render an empty table
      return res.status(200).json({
        success: true,
        data: {
          products: [],
          premixed: [],
          obm: [],
          bulkTankSetupFee: "",
          taxRate: "",
          applyPricesOption: "To All",
          fromDate: "",
        },
      });
    }

    res.status(200).json({ success: true, data: snapshot });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

// ─────────────────────────────────────────────────────────────────
// POST  /api/ug-inventory/:wellId
// Upsert (create or replace) the full snapshot for a well
// ─────────────────────────────────────────────────────────────────
export const saveUgInventory = async (req, res) => {
  try {
    const { wellId } = req.params;
    const {
      products,
      premixed,
      obm,
      bulkTankSetupFee,
      taxRate,
      applyPricesOption,
      fromDate,
    } = req.body;

    const snapshot = await UgInventorySnapshot.findOneAndUpdate(
      { wellId },
      {
        $set: {
          products:          products          ?? [],
          premixed:          premixed          ?? [],
          obm:               obm               ?? [],
          bulkTankSetupFee:  bulkTankSetupFee  ?? "",
          taxRate:           taxRate           ?? "",
          applyPricesOption: applyPricesOption ?? "To All",
          fromDate:          fromDate          ?? "",
        },
      },
      { new: true, upsert: true, runValidators: true }
    );

    res.status(200).json({
      success: true,
      message: "Inventory saved successfully",
      data: snapshot,
    });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

File: MudPro-Desktop-App/lib/modules/UG/right_pannel/inventory/controller/ug_inventory_product_controller.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:mudpro_desktop_app/api_endpoint/api_endpoint.dart';
import 'package:mudpro_desktop_app/modules/UG/model/inventory_model.dart';
import 'package:mudpro_desktop_app/modules/UG/right_pannel/inventory/model/ug_inventory_product_model.dart';
import 'package:mudpro_desktop_app/modules/company_setup/model/service_model.dart';

class InventoryProductsService {
  // ✅ Change to your actual backend URL
  static const String _baseUrl = ApiEndpoint.baseUrl;

  // ─── Apply: Save all categories at once ─────────────────
  static Future<void> applyInventoryData({
    required String wellId,
    required List<ProductInventoryModel> products,
    required List<PremixModel> premixed,
    required List<ObmModel> obm,
    List<PackageItem>? packages,
    List<ServiceItem>? services,
    List<EngineeringItem>? engineering,
    String? bulkTankSetupFee,
    String? taxRate,
    String? applyPricesOption,
    String? fromDate,
  }) async {
    final url = '${_baseUrl}ug-inventory/$wellId';
    print('🔵 [POST] Saving inventory to URL: $url');
    final uri = Uri.parse(url);

    final bodyData = {
      'wellId': wellId,
      'products': products.map((ProductInventoryModel p) => p.toJson()).toList(),
      'premixed': premixed.map((PremixModel p) => p.toJson()).toList(),
      'obm': obm.map((ObmModel o) => o.toJson()).toList(),
      'packages': packages?.map((e) => e.toJson()).toList() ?? [],
      'services': services?.map((e) => e.toJson()).toList() ?? [],
      'engineering': engineering?.map((e) => e.toJson()).toList() ?? [],
    };

    if (bulkTankSetupFee != null) bodyData['bulkTankSetupFee'] = bulkTankSetupFee;
    if (taxRate != null) bodyData['taxRate'] = taxRate;
    if (applyPricesOption != null) bodyData['applyPricesOption'] = applyPricesOption;
    if (fromDate != null) bodyData['fromDate'] = fromDate;

    final body = jsonEncode(bodyData);

    final response = await http.post(
      uri,
      headers: ApiEndpoint.jsonHeaders,
      body: body,
    );

    print('🟢 [POST] statusCode: ${response.statusCode}');
    print('🟢 [POST] responseBody: ${response.body}');

    final json = jsonDecode(response.body);

    if (response.statusCode != 200 || json['success'] != true) {
      throw Exception(json['message'] ?? 'Failed to save inventory');
    }
  }

  // ─── Get: Fetch full snapshot ───────────────────────────────
  static Future<Map<String, dynamic>> getInventoryData(String wellId) async {
    final url = '${_baseUrl}ug-inventory/$wellId';
    print('🔵 [GET] Fetching inventory from URL: $url');
    final uri = Uri.parse(url);

    final response = await http.get(uri, headers: ApiEndpoint.jsonHeaders);

    print('🟢 [GET] statusCode: ${response.statusCode}');
    print('🟢 [GET] responseBody: ${response.body}');

    final json = jsonDecode(response.body);

    if (response.statusCode != 200 || json['success'] != true) {
      throw Exception(json['message'] ?? 'Failed to fetch inventory');
    }

    return json['data'];
  }

  // ─── Specific Fetch Methods ─────────────────────────────────

  static Future<List<ProductInventoryModel>> fetchProducts(String wellId) async {
    final url = '${_baseUrl}ug-inventory/products/$wellId';
    print('🔵 [GET] Fetching products from URL: $url');
    final response = await http.get(Uri.parse(url), headers: ApiEndpoint.jsonHeaders);
    print('🟢 [GET] statusCode: ${response.statusCode}');
    print('🟢 [GET] responseBody: ${response.body}');
    final json = jsonDecode(response.body);
    if (response.statusCode == 200 && json['success'] == true) {
      return (json['data'] as List).map((e) => ProductInventoryModel.fromJson(e)).toList();
    }
    return [];
  }

  static Future<List<PackageItem>> fetchPackages(String wellId) async {
    final url = '${_baseUrl}ug-inventory/packages/$wellId';
    print('🔵 [GET] Fetching packages from URL: $url');
    final response = await http.get(Uri.parse(url), headers: ApiEndpoint.jsonHeaders);
    final json = jsonDecode(response.body);
    if (response.statusCode == 200 && json['success'] == true) {
      return (json['data'] as List).map((e) => PackageItem.fromJson(e)).toList();
    }
    return [];
  }

  static Future<List<ServiceItem>> fetchServices(String wellId) async {
    final url = '${_baseUrl}ug-inventory/services/$wellId';
    print('🔵 [GET] Fetching services from URL: $url');
    final response = await http.get(Uri.parse(url), headers: ApiEndpoint.jsonHeaders);
    final json = jsonDecode(response.body);
    if (response.statusCode == 200 && json['success'] == true) {
      return (json['data'] as List).map((e) => ServiceItem.fromJson(e)).toList();
    }
    return [];
  }

  static Future<List<EngineeringItem>> fetchEngineering(String wellId) async {
    final url = '${_baseUrl}ug-inventory/engineering/$wellId';
    print('🔵 [GET] Fetching engineering from URL: $url');
    final response = await http.get(Uri.parse(url), headers: ApiEndpoint.jsonHeaders);
    final json = jsonDecode(response.body);
    if (response.statusCode == 200 && json['success'] == true) {
      return (json['data'] as List).map((e) => EngineeringItem.fromJson(e)).toList();
    }
    return [];
  }
}
